src/OpenTelemetry/Logs/Pool/LogRecordSharedPool.cs

```diff
@@ -0,0 +1,172 @@
+    {
+        internal const int DefaultMaxPoolSize = 2048;
+        internal const int DefaultMaxNumberOfAttributes = 64;
+        internal const int DefaultMaxNumberOfScopes = 16;
+
+        internal static LogRecordSharedPool Current = new(DefaultMaxPoolSize);
+
+        internal readonly int Capacity;
+        private readonly LogRecord?[] pool;
+        private long rentIndex;
+        private long returnIndex;
+
+        internal LogRecordSharedPool(int capacity)
+        {
+            this.Capacity = capacity;
+            this.pool = new LogRecord?[capacity];
+        }
+
+        internal int Count => (int)(Volatile.Read(ref this.returnIndex) - Volatile.Read(ref this.rentIndex));
+
+        public LogRecord Rent()
+        {
+            while (true)
+            {
+                var rentSnapshot = Volatile.Read(ref this.rentIndex);
+                var returnSnapshot = Volatile.Read(ref this.returnIndex);
+
+                if (rentSnapshot >= returnSnapshot)
+                {
+                    break; // buffer is empty
+                }
+
+                if (Interlocked.CompareExchange(ref this.rentIndex, rentSnapshot + 1, rentSnapshot) == rentSnapshot)
+                {
+                    var logRecord = Interlocked.Exchange(ref this.pool[rentSnapshot % this.Capacity], null);
+                    if (logRecord == null && !this.TryRentCoreRare(rentSnapshot, out logRecord))
+                    {
+                        continue;
+                    }
+
+                    logRecord.ResetReferenceCount();
+                    return logRecord;
+                }
+            }
+
+            var newLogRecord = new LogRecord();
+            newLogRecord.ResetReferenceCount();
+            return newLogRecord;
+        }
+
+        public void Return(LogRecord logRecord)
+        {
+            if (logRecord.RemoveReference() != 0)
+            {
+                return;
+            }
+
+            Clear(logRecord);
+
+            while (true)
+            {
+                var rentSnapshot = Volatile.Read(ref this.rentIndex);
+                var returnSnapshot = Volatile.Read(ref this.returnIndex);
+
+                if (returnSnapshot - rentSnapshot >= this.Capacity)
+                {
+                    return; // buffer is full
+                }
+
+                if (Interlocked.CompareExchange(ref this.returnIndex, returnSnapshot + 1, returnSnapshot) == returnSnapshot)
+                {
+                    // If many threads are hammering rent/return it is possible
+                    // for two threads to write to the same index. In that case
+                    // only one of the logRecords will make it back into the
+                    // pool. Anything lost in the race will collected by the GC
+                    // and the pool will issue new instances as needed
+                    this.pool[returnSnapshot % this.Capacity] = logRecord;
```

Comment:
I like this idea, however I think it might be optimizing for the wrong case. The race called out in the comment is for "many threads hammering rent/return" which is more of a theoretic exercise. The actual use (right now) is many threads renting/reading (log emitters), one thread returning/writing (batch exporter). In that scenario, this race should really never be hit. So if we put in the `Interlocked.CompareExchange` that is more expensive than just doing `this.pool[returnSnapshot % this.Capacity] = logRecord` which should work for actual use case we have right now. Are you open to me updating the comment with the solution and considerations so we have it captured (in the event the use case changes) but we leave as-is for now so we have better perf for exporters returning records?

Reply:
Just updated the comment with more details.